### PR TITLE
downsample: fix stale handling, improve chunk count heuristic

### LINKED ISSUE #197: Small and unbalanced downsampled chunks produced
I'm currently debugging a case where a downsampled series is producing a lot of small chunks rather than few sufficiently large ones. A small fraction of series are also finalized with chunks with a timestamp range of `[-1, -1]` for which there's no logical reason in the base data.

This does likely not affect correctness of query results but was caught by the new validations procedures added in #194. It's probably really worth keeping those around forever to catch regressions into behavior like this early.

### PR DESCRIPTION
**Pull request by fabxc**
Fixes #197

### CODE REVIEW

**Comment by fabxc** [comment]
Added one last fix to prevent single-sample overlaps between downsampling of already downsampled chunks.

**Comment by fabxc** [comment]
@Bplotka I removed the per-compaction group metrics – explanation in commit message. I hope you don't mind. Overall I think they didn't provide much value as metrics.

File: pkg/compact/downsample/downsample.go [hunk 1]
```diff
@@ -361,27 +365,56 @@ func currentWindow(t, r int64) int64 {
...
+// targetChunkCount calculates how many chunks should be produced when downsampling a series.
+// It consider the total time range, the number of input sample, the input and output resolution.
+func targetChunkCount(mint, maxt, inRes, outRes int64, count int) (x int) {
+	// We compute how many samples we could produce for the given time range and adjust
+	// it by how densely the range is actually filled given the number of input samples and their
+	// resolution.
+	maxSamples := float64((maxt - mint) / outRes)
+	expSamples := int(maxSamples*rangeFullness(mint, maxt, inRes, count)) + 1
+
+	// Increase the number of target chunks until we each chunk will have less than
```
Comment (Bplotka) [inline pkg/compact/downsample/downsample.go]: `s/until we each/until each`

File: pkg/compact/downsample/downsample.go [hunk 2]
```diff
@@ -431,15 +471,18 @@ func downsampleBatch(data []sample, resolution int64, add func(int64, *aggregato
 }
 
 // downsampleAggr downsamples a sequence of aggregation chunks to the given resolution.
-func downsampleAggr(chks []AggrChunk, buf *[]sample, mint, maxt, resolution int64) ([]chunks.Meta, error) {
+func downsampleAggr(chks []AggrChunk, buf *[]sample, mint, maxt, inResolution, outResolution int64) ([]chunks.Meta, error) {
```
Comment (Bplotka) [inline pkg/compact/downsample/downsample.go]: in consistent naming. In `targetChunkCount` you have `inRes` and `outRes`. Please choose one style. OK with both.

**Comment by fabxc** [comment]
Rebased and comments fixed.

**Pull request merged**